broker/night_conductor/end_night/process_pubsub_counters.py

```diff
@@ -0,0 +1,386 @@
+
+        self.metadata_dicts_list.append(metadata)
+
+        # for now, we assume everything went fine... update this if problems arise
+        success = True
+        return success
+
+    def _stash_dicts_to_json_file(self):
+        # for debugging
+        fname = f"metadata/{self.subscription}.json"
+        with open(fname, "w") as fout:
+            json.dump(self.metadata_dicts_list, fout, allow_nan=True)
+
+    def _package_metadata_into_df(self):
+        # cast to a dataframe and set the index and columns
+        oid = self.schema_map["objectId"]  # survey-specific field name for the objectId
+        sid = self.schema_map["sourceId"]  # survey-specific field name for the sourceId
+
+        # cast the collected metadata to a df
+        df = pd.DataFrame(self.metadata_dicts_list)
+
+        # set the index with explicit types
+        if (oid in df.columns) and (sid in df.columns):
+            df = df.astype({oid: str, sid: str})
+            df = df.set_index([oid, sid])
+        else:
+            # currently this is just the "alerts" stream
+            # and is handled in MetadataCollector._add_ids_to_alerts_stream()
+            df = df.astype({"message_id": int}).set_index("message_id")
+
+        # attach topic name stub to column names, separated by double underscore
+        df.rename(columns=lambda x: f"{x}__{self.topic_stub}", inplace=True)
+
+        # format column names; BigQuery accepts only letters, numbers, and underscores
+        df.rename(columns=lambda x: x.replace("-", "_"), inplace=True)
+        df.rename(columns=lambda x: x.replace(".", "_"), inplace=True)
+
+        self.metadata_df = df
+
+    def _package_publish_time(self, publish_time):
+        # publish_time timestamp is given as two ints: seconds and nanoseconds
+        # package it into a single number
+
+        # preserve zero padding
+        tmpnanos = f"{publish_time.nanos*1e-9}".split(".")[1][:9]
+
+        # concat. use string type to ensure exact representation
+        pubtime = f"{publish_time.seconds}.{tmpnanos}"
+
+        return pubtime
+
+    def _package_alert_avros_metadata(self, msg_data, msg_metadata):
+        # clarify the metadata names
+        # extract *file* metadata stored in the message body
+
+        # rename field for clarity and to avoid name collision with DIA objectId
+        msg_metadata["filename"] = msg_metadata.pop("objectId")
+
+        # extract the file's custom metadata set in the ps_to_gcs cloud fnc
+        file_metadata = json.loads(msg_data.decode("UTF-8"))
+        file_custom_metadata = file_metadata["metadata"]
+
+        metadata = {**msg_metadata, **file_custom_metadata}
+
+        return metadata
+
+
+if __name__ == "__main__":  # noqa
+    parser = argparse.ArgumentParser()
+    # survey
+    parser.add_argument(
+        "--survey",
+        dest="survey",
+        default="ztf",
+        help="The broker instance's `survey` keyword. Names the resources.\n",
```

Comment:
"Broker instance `survey` keyword.  Used to label resources."

Might be more clear.  (similar below for `--testid`).